### ISSUE TITLE
Web/mWeb- Sign in - Endless loading spinner appears when re login

### ISSUE THREAD

**Issue by kavimuru** [Expensify/Expensify.cash#3488]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Expected Result:
Able to log in to the account without any issue

## Actual Result:
Infinite loading spinner appears on tapping "continue" after entering the email
## Action Performed:
1. Go to https://staging.expensify.cash/
2. Log in with any account
3. Log out
4. Log in again with same account or different account
## Workaround:
Unknown

## Platform:

 Web ✔️
 iOS
 Android 
 Desktop App
 Mobile Web ✔️

**Version Number:** 
1.0.66-0
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/121415171-6f23d100-c935-11eb-9f7d-4b83af2584e2.mp4

**Expensify/Expensify Issue URL:**

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @Luke9389 (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by Christinadobrzyn** [comment]
Testing this in Staging v1.0.55-1

I'm not getting the same loading page when I sign in to a different account, I assume that is because I'm not logging into the account as fast as in the video @kavimuru so kindly created. 

I don't see any other GH's about this issue so I think it's up to engineering if this is something to fix now or wait.

**Comment by trjExpensify** [comment]
👋 @Christinadobrzyn, did you try to reproduce on Web and mWeb and on version `v1.0.66-0` or higher? 

All deploy blockers are set to `hourly` for urgency. Let's keep it on that until we prove this isn't an issue or not. Essentially, if this gets deployed to production and users can't sign-in to E.cash, that's pretty bad. 

FWIW, I can see the issue on Web, and now I'm unable to sign-in 😅 

![image](https://user-images.githubusercontent.com/16232057/121448241-1f71f380-c98f-11eb-83a5-9d3ed64884a7.png)

**Comment by Christinadobrzyn** [comment]
Ah this is another Alpha staging situation - I don't have a version higher than v1.0.55-1. Asking permission for v1.0.66-0 in [Slack](https://expensify.slack.com/archives/C03U7DCU4/p1623265829213700).

**Comment by trjExpensify** [comment]
Yeah, you'll need the alpha for the latest staging version on the mobile app. For testing Web/mWeb issues like this one, if you go to staging.expensify.cash, those will always be the latest for yah.

>Does this fix need an external contractor or are we fixing this internally?

hourly deploy blockers are typically kept internal 👍

**Comment by trjExpensify** [comment]
@Luke9389 is largely OoO today, so we're repulling the deploy blocker assignment trigger given that it's an hourly deploy blocker that's blocking N5.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @jasperhuangg (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by Julesssss** [comment]
As [Tom mentioned](https://github.com/Expensify/Expensify.cash/issues/3488#issuecomment-858778535), we need to re-roll to find someone who can take a look at this today.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @pecanoro (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by trjExpensify** [comment]
Bingo! :)

**Comment by pecanoro** [comment]
Sorry guys, but I definetely can't take care either, cash bonus payroll project is extremely urgent and have a super tight deadline 😢

**Comment by pecanoro** [comment]
Can we use the demolition label here? It has really clear reproduction steps

**Comment by MelvinBot** [comment]
@trjExpensify the `Demolition` label can only be added by an Engineer or a member of the Engineering Demolition Team.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @alex-mechler (`Demolition`), see https://stackoverflow.com/c/expensify/questions/8099 for more details.

**Comment by trjExpensify** [comment]
Yep, cool. As long as someone takes it, as it is preventing us from being able to ship N5. 👍

**Comment by alex-mechler** [comment]
Okay, found the root cause. When we go to sign out, we clean up the logins we create, and [delete it here](https://github.com/Expensify/Expensify.cash/blob/655ba416d552d5c88e57977a6e0165fb7eb7ab58/src/libs/actions/Session.js#L69). However, since we have sped up onyx, we now delete our auth token before we fire this request, causing it to fail, and [lock the network](https://github.com/Expensify/Expensify.cash/blob/6cf1a56df670a11bf61aa67eeb64c1f87161dea1/src/libs/API.js#L61) since we don't have a auth token. We only unlock the network when we get a valid authtoken again. 

Moving the delete login request before we clear the authtoken solves this problem, but I am running into a few other problems that still have to be solved with it

**Comment by roryabraham** [comment]
Oops, @kidroca made Onyx _too_ fast 🙃

**Comment by kidroca** [comment]
I think we have the same problem we concluded here: https://github.com/Expensify/Expensify.cash/issues/3485#issuecomment-858705481

And maybe it would be best to move the call the `keysChanged` to the next tick as suggested in the link above
This way you won't have to change the code here and on other places suffering from this speed up

cc @marcaaron

**Comment by alex-mechler** [comment]
Alright, so I have a [fix PR](https://github.com/Expensify/Expensify.cash/pull/3535). It has uncovered some other bugs, which I think are related to the cache that was recently added.

1. When signing into an account after the signout, the loading spinner is _not_ cleared on the password page. `account.loading` is set to false in local storage at this time. 
2. After signing into the account after the signout, You are brought to `/r/`, with no report ID. 

Both of these issues are solved when using the commit before caching was introduced to Onyx, `0cdf133867dccbd0b99090850130fef3442e826f`, which leads me to believe the caching is causing these issues.

**Comment by Julesssss** [comment]
> And maybe it would be best to make a PR to move the call to `keysChanged` to the next tick as suggested in the link above
> This way you won't have to change the code here and on other places suffering from this speed up

@kidroca maybe this is the best path forward, would you be able to raise a PR with this change so we can discuss it further? Do you think that would resolve the issues raised by Alex above?

**Comment by Julesssss** [comment]
> I tested this locally, and while it did work, it introduced another issue where the Magic Sign link page would briefly appear on the screen before switching over to the Password Form.

Hey @alex-mechler. It looks like this might be the best path forward to unblocking the N5 release. Would you mind pushing this change so we can investigate a fix to the hopefully final issue of the Magic sign in appearing? Thanks 🙇‍♂️

**Comment by alex-mechler** [comment]
Sure, I can take care of that

**Comment by alex-mechler** [comment]
So with that fix in Onyx, issue 2 (After signing into the account after the signout, You are brought to /r/, with no report ID) still exists after signing out and then back into a new account. `currentlyViewedReportID` is `NaN` in Local Storage.

**Comment by Julesssss** [comment]
PR looks good to me. As discussed, I think the remaining redirect to `/r` issue is probably not worth blocking the release over.

**Comment by alex-mechler** [comment]
PR is up, follow up issue for the `/r/` [bug here](https://github.com/Expensify/Expensify/issues/167034)
